keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,883 @@
+- The implementation must handle image pull secrets and other registry authentication mechanisms.
+- Performance considerations must be taken into account, especially for large images or artifacts.
+
+### Vocabulary: OCI Images, Artifacts, and Objects
+
+**OCI Image:**
+   - A container image that conforms to the Open Container Initiative (OCI) Image Specification.
+     It includes a filesystem bundle and metadata required to run a container.
+   - Consists of multiple layers (each layer being a tarball), a manifest (which lists the layers), and a config file
+     (which provides configuration data such as environment variables, entry points, etc.).
+   - **Use Case:** Used primarily for packaging and distributing containerized applications.
+
+**OCI Artifact:**
+   - An artifact describes any content that is stored and distributed using the OCI image format.
+     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
+   - Artifacts use the same image manifest and layer structure but may contain different types of data
+     within those layers. The artifact manifest can have media types that differ from those in standard container images.
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Ensure compatibility with existing features.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet, Scheduler)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Enhancements to the scheduler to recognize and properly allocate resources for this new volume type.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps for:
+  - Determining the file location on the host filesystem.
+  - Managing garbage collection of unused images and artifacts.
+  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.
```

Comment:
are these changes on the kubelet? seems related to @samuelkarp question on https://github.com/kubernetes/enhancements/pull/4642/files#r1612384741

Reply:
Yes I added more detail about CRI and kubelet changes.